core/src/main/java/com/graphhopper/routing/ch/CHAlgoFactoryDecorator.java

```diff
@@ -279,38 +278,41 @@ public void setPreparationThreads(int preparationThreads) {
     }
 
     public void prepare(final StorableProperties properties) {
+        ExecutorCompletionService completionService = new ExecutorCompletionService<>(threadPool);
         int counter = 0;
         for (final PrepareContractionHierarchies prepare : getPreparations()) {
             LOGGER.info((++counter) + "/" + getPreparations().size() + " calling CH prepare.doWork for " + prepare.getWeighting() + " ... (" + Helper.getMemInfo() + ")");
             final String name = AbstractWeighting.weightingToFileName(prepare.getWeighting());
-            threadPool.execute(new Runnable() {
+            completionService.submit(new Runnable() {
                 @Override
                 public void run() {
                     String errorKey = CH.PREPARE + "error." + name;
                     try {
                         // toString is not taken into account so we need to cheat, see http://stackoverflow.com/q/6113746/194609 for other options
                         Thread.currentThread().setName(name);
-
                         properties.put(errorKey, "CH preparation incomplete");
                         prepare.doWork();
                         properties.remove(errorKey);
                         properties.put(CH.PREPARE + "date." + name, Helper.createFormatter().format(new Date()));
                     } catch (Exception ex) {
                         LOGGER.error("Problem while CH preparation " + name, ex);
                         properties.put(errorKey, ex.getMessage());
+                        throw ex;
                     }
                 }
-            });
+            }, name);
```

Comment:
After the thread finishes the Future returns this variable as result. We could write any result that we want, I just figured the name would be nice and could be usable in the future. For example when retrieving the result we could state: "Finished Thread: "+name. This is different from the Thread name.